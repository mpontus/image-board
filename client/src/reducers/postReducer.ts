import { Reducer } from "redux";
import {
  Action,
  CREATE_POST,
  CREATE_POST_REJECT,
  CREATE_POST_RESOLVE,
  DELETE_POST,
  DELETE_POST_REJECT,
  DELETE_POST_RESOLVE,
  LIKE_POST,
  LIKE_POST_REJECT,
  LOAD_POSTS_RESOLVE
} from "../actions";
import { PostData, Progress } from "../models";

export interface State {
  readonly loading: boolean;
  readonly lastPage: number | null;
  readonly total: number | null;
  readonly ids: ReadonlyArray<string>;
  readonly byId: Readonly<{ [id: string]: PostData | undefined }>;
  readonly pendingIds: ReadonlyArray<string>;
  readonly isPendingById: Readonly<{ [id: string]: boolean }>;
  readonly isDeletedById: Readonly<{ [id: string]: boolean }>;
  readonly instances: Readonly<{ [id: string]: string }>;
  readonly progress: Readonly<{ [id: string]: Progress }>;
}

const initialState: State = {
  loading: true,
  lastPage: null,
  total: null,
  ids: [],
  byId: {},
  pendingIds: [],
  isPendingById: {},
  isDeletedById: {},
  instances: {},
  progress: {}
};

const reducer: Reducer<State, Action> = (
  state = initialState,
  action: Action
) => {
  switch (action.type) {
    case LOAD_POSTS_RESOLVE: {
      const { page, total, posts } = action.payload;

      return {
        ...state,
        loading: false,
        lastPage: page,
        total,
        ids: state.ids.concat(posts.map(post => post.id)),
        byId: posts.reduce(
          (acc, post) => ({
            ...acc,
            [post.id]: post
          }),
          state.byId
        )
      };
    }

    case CREATE_POST: {
      const { post } = action.payload;

      return {
        ...state,
        pendingIds: [post.id, ...state.pendingIds],
        isPendingById: {
          ...state.isPendingById,
          [post.id]: true
        },
        byId: {
          ...state.byId,
          [post.id]: post
        }
      };
    }

    case CREATE_POST_RESOLVE: {
      const { post, instance } = action.payload;

      return {
        ...state,
        pendingIds: state.pendingIds.filter(id => id !== post.id),
        ids: [post.id, ...state.ids],
        instances: {
          ...state.instances,
          [post.id]: instance.id
        },
        byId: {
          ...state.byId,
          [instance.id]: instance
        }
      };
    }

    case CREATE_POST_REJECT: {
      const { post } = action.payload;

      return {
        ...state,
        pendingIds: state.pendingIds.filter(id => id !== post.id),
        byId: {
          ...state.byId,
          [post.id]: undefined
        }
      };
    }

    case DELETE_POST: {
      const { post } = action.payload;

      return {
        ...state,
        isDeletedById: {
          ...state.isDeletedById,
          [post.id]: true
        }
      };
    }

    case DELETE_POST_RESOLVE: {
      const { post } = action.payload;

      return {
        ...state,
        byId: {
          ...state.byId,
          [post.id]: undefined
        }
      };
    }

    case DELETE_POST_REJECT: {
      const { post } = action.payload;

      return {
        ...state,
        isDeletedById: {
          ...state.isDeletedById,
          [post.id]: false
        }
      };
    }

    case LIKE_POST: {
      const { post, value } = action.payload;
      const current = state.byId[post.id];

      if (!current) {
        return state;
      }

      return {
        ...state,
        byId: {
          ...state.byId,
          [post.id]: {
            ...current,
            isLiked: value > 0,
            likesCount: current.likesCount + value
          }
        }
      };
    }

    case LIKE_POST_REJECT: {
      const { post, value } = action.payload;
      const current = state.byId[post.id];

      if (!current) {
        return state;
      }

      return {
        ...state,
        byId: {
          ...state.byId,
          [post.id]: {
            ...current,
            isLiked: !(value > 0),
            likesCount: current.likesCount - value
          }
        }
      };
    }

    default:
      return state;
  }
};

export default reducer;
