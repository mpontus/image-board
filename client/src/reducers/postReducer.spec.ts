import {
  Action,
  CREATE_POST,
  loadPostsResolve,
  createPostResolve,
  createPostReject
} from "../actions";
import postReducer from "./postReducer";

const postFixtures = [
  {
    id: "1",
    picture: {
      url:
        "https://images.unsplash.com/photo-1535412833400-85426926b8c1?ixlib=rb-0.3.5&ixid=eyJhcHBfaWQiOjEyMDd9&s=47e717c53dae51ed4a300fffa13733a8&auto=format&fit=crop&w=500&q=60",
      width: 500,
      height: 333
    },
    author: {
      id: "auth9|123123",
      name: "Foo bar",
      avatarUrl:
        "https://images.unsplash.com/profile-1532310311737-e56bb5caa506?dpr=1&auto=format&fit=crop&w=64&h=64&q=60&crop=faces&bg=fff"
    },
    likesCount: 1,
    isLiked: true,
    timestamp: 1535731213512
  },
  {
    id: "2",
    picture: {
      url:
        "https://images.unsplash.com/photo-1535406110845-88cbe4f661bc?ixlib=rb-0.3.5&ixid=eyJhcHBfaWQiOjEyMDd9&s=f4a0c1c0797ffc0e0008c764bf37fdf4&auto=format&fit=crop&w=500&q=60",
      width: 500,
      height: 281
    },
    author: {
      id: "auth9|123123",
      name: "Foo bar",
      avatarUrl:
        "https://images.unsplash.com/profile-1532310311737-e56bb5caa506?dpr=1&auto=format&fit=crop&w=64&h=64&q=60&crop=faces&bg=fff"
    },
    likesCount: 1,
    isLiked: true,
    timestamp: 1535731213512
  }
];

const initialState = postReducer(undefined, {} as any);

describe("Post reducer", () => {
  describe("initial state", () => {
    it("should match snapshot", () => {
      expect(initialState).toMatchSnapshot();
    });
  });

  describe("listing retrieval", () => {
    it("should add posts to the state", () => {
      const action = loadPostsResolve({
        total: 2,
        items: postFixtures
      });

      const state = postReducer(initialState, action);

      expect(state).toEqual({
        ...initialState,
        loading: false,
        total: 2,
        ids: [postFixtures[0].id, postFixtures[1].id],
        byId: {
          [postFixtures[0].id]: postFixtures[0],
          [postFixtures[1].id]: postFixtures[1]
        }
      });
    });
  });

  describe("post creation", () => {
    it("should add the post optimistically", () => {
      const file = new File([""], "picture.png");
      const action: Action = {
        type: CREATE_POST,
        payload: {
          file,
          post: postFixtures[0]
        }
      };

      const state = postReducer(initialState, action);

      expect(state).toEqual({
        ...initialState,
        pendingIds: [postFixtures[0].id],
        byId: {
          [postFixtures[0].id]: postFixtures[0]
        }
      });
    });

    it("should remove post from pending after it's committed", () => {
      const action = createPostResolve(postFixtures[0], postFixtures[1]);
      const state0 = {
        ...initialState,
        pendingIds: [postFixtures[0].id],
        byId: {
          [postFixtures[0].id]: postFixtures[0]
        }
      };
      const state1 = postReducer(state0, action);

      expect(state1).toEqual({
        ...state0,
        pendingIds: [],
        ids: [postFixtures[0].id],
        byId: {
          ...state0.byId,
          [postFixtures[1].id]: postFixtures[1]
        },
        instances: {
          [postFixtures[0].id]: postFixtures[1].id
        }
      });
    });

    it("should remove post when creation is unsuccessful", () => {
      const action = createPostReject(postFixtures[0], new Error("foo"));
      const state0 = {
        ...initialState,
        pendingIds: [postFixtures[0].id],
        byId: {
          [postFixtures[0].id]: postFixtures[0]
        }
      };
      const state1 = postReducer(state0, action);

      expect(state1).toEqual({
        ...state0,
        pendingIds: [],
        byId: {}
      });
    });
  });
});
