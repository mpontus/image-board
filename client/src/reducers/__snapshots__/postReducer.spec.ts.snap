// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post reducer initial state should match snapshot 1`] = `
Object {
  "byId": Object {},
  "ids": Array [],
  "instances": Object {},
  "loading": false,
  "pendingIds": Array [],
  "total": null,
}
`;
